File: Bachelor/Semester4/Systems_for_Design_and_Implementation/C#/ProiectCShMPP/Networking/Utils/AbstractServer.cs
using System;
using System.Net;
using System.Net.Sockets;
using log4net;

namespace Networking.Utils
{
    public abstract class AbstractServer
    {
        private TcpListener _server;
        private readonly int _port;
        private readonly string _ip;
        private static readonly ILog _logger = LogManager.GetLogger(typeof(AbstractServer));

        public AbstractServer(string ip, int port)
        {
            this._port = port;
            this._ip = ip;
        }

        public void Start()
        {
            _logger.InfoFormat("Entry Start Server");
            _server = new TcpListener(IPAddress.Parse(_ip), _port);
            _server.Start();
            while (true)
            {
                Console.WriteLine("Waiting clients...");
                TcpClient client = _server.AcceptTcpClient();
                Console.WriteLine("{0} connected", client);
                _logger.InfoFormat("{0} connected", client);
                ProcessRequest(client);
            }
        }

        protected abstract void ProcessRequest(TcpClient client);
    }
}

File: Bachelor/Semester5/Parallel_and_Distributed_Programming/P1/Networking/src/main/java/utils/AbstractServer.java
package utils;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.TimeUnit;

public abstract class AbstractServer {
    private ServerSocket server;
    private final int port;
    private Timer timer;
    private static final int MINUTES_TO_RUN = 2;

    protected AbstractServer(int port) {
        this.port = port;
    }

    public void start() {
        scheduleTimeToCloseServer();
        scheduleChecks();
        try {
            server = new ServerSocket(port);
            while (true) {
                System.out.println("Waiting for clients");
                Socket client = server.accept();
                System.out.println("Somebody connected !");
                processRequest(client);
            }
        } catch (Exception e) {
            System.out.println("CLOSED");
        }
    }

    protected abstract void processRequest(Socket client);

    protected abstract void scheduleChecks();

    public void scheduleTimeToCloseServer() {
        timer = new Timer();
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                stop();
                timer.cancel();
                timer.purge();
            }
        }, TimeUnit.MINUTES.toMillis(MINUTES_TO_RUN));
    }

    public void stop() {
        try {
            server.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
